src/coreclr/src/vm/callcounting.cpp

```diff
@@ -0,0 +1,1735 @@
+                callCountingInfo->SetStage(CallCountingInfo::Stage::Complete);
+                return true;
+            } while (false);
+
+            callCountingStub = callCountingInfo->GetCallCountingStub();
+            if (callCountingStub != nullptr)
+            {
+                break;
+            }
+        }
+        else
+        {
+            _ASSERTE(activeCodeVersion.GetOptimizationTier() == NativeCodeVersion::OptimizationTier0);
+
+            // If the tiering delay is active, postpone further work
+            if (GetAppDomain()
+                    ->GetTieredCompilationManager()
+                    ->TrySetCodeEntryPointAndRecordMethodForCallCounting(methodDesc, codeEntryPoint))
+            {
+                return true;
+            }
+
+            CallCount callCountThreshold = (CallCount)g_pConfig->TieredCompilation_CallCountThreshold();
+            _ASSERTE(callCountThreshold != 0);
+            _ASSERTE(callCountThreshold <= MaximumCallCountThreshold);
+
+            NewHolder<CallCountingInfo> callCountingInfoHolder = new CallCountingInfo(activeCodeVersion, callCountThreshold);
+            callCountingManager->m_callCountingInfoByCodeVersionHash.Add(callCountingInfoHolder);
+            callCountingInfo = callCountingInfoHolder.Extract();
+        }
+
+        if (!g_pConfig->TieredCompilation_UseCallCountingStubs())
+        {
+            // Call counting is not yet complete, so reset or don't set the code entry point to continue counting calls
+
+            if (wasMethodCalled)
+            {
+                return false;
+            }
+
+            // This path is reached after activating a code version when publishing its code entry point. The method may
+            // currently be pointing to the code entry point of a different code version, so an explicit reset is necessary.
+            methodDesc->ResetCodeEntryPoint();
+            return true;
+        }
+
+        callCountingStub =
+            callCountingManager->m_callCountingStubAllocator.AllocateStub(
+                callCountingInfo->GetRemainingCallCountCell(),
+                codeEntryPoint);
+        callCountingInfo->SetCallCountingStub(callCountingStub);
+    } while (false);
+
+    PCODE callCountingCodeEntryPoint = callCountingStub->GetEntryPoint();
+    if (methodDesc->MayHaveEntryPointSlotsToBackpatch())
+    {
+        // The call counting stub should not be the entry point that is called first in the process of a call
+        // - Stubs should be deletable. Many methods will have call counting stubs associated with them, and although the memory
+        //   involved is typically insignificant compared to the average memory overhead per method, by steady-state it would
+        //   otherwise be unnecessary memory overhead serving no purpose.
+        // - In order to be able to delete a stub, the jitted code of a method cannot be allowed to load the stub as the entry
```

Comment:
I had discussed this with Sergiy, I'll post some numbers separately. Most of the noticeable overhead from deleting stubs seems to be in restoring some of them through the prestub, that's the main reason for the 4 K minimum completed stubs before deleting, and perhaps it could be a lazier heuristic than that, anyway I believe that's the main risk factor. It seems to be working decently for now.